### PR TITLE
Codebase refactoring

### PR DESCRIPTION
**Pull request by rxin**
I refactored the code substantially, and now all the Vertex partition specific implementation is encapsulated in VertexPartition, and GraphImpl / VertexSetRDD (renamed VertexRDD) don't operate on bitmaps anymore.

Also added an EdgeRDD[ED] class to replace the previous RDD[Edge[ED]].

### CODE REVIEW

**Comment by jegonzal** [comment]
I suspect the test are failing:

   org.apache.spark.SparkException: can't zip join VertexSetRDDs with different indexes

because we are testing equivalence of IndexRDDs which was previously true when both VertexSetRDDs contained a reference to the same IndexRDD but now that the IndexRDD is implicit two VertexSetRDDs which are composed of the same internal HashSet for each partition will actually have different IndexRDDs.

**Comment by ankurdave** [comment]
Good news: this has no significant impact on performance.

| Commit | Runtime |
| --- | --- |
| 088995f917548ac549397f24916ea72b0c3fc9d0 (master) | 261.6 +/- 7.13 s |
| 8701cb55e6590bd9069f09910a54968ba5941804 (clean1) | 259.3 +/- 6.02 s |

**Comment by jegonzal** [comment]
Wait, is that good news? :-)

—

**Comment by rxin** [comment]
Yes. This wasn't meant for perf improvement. Perf coming next.

File: graph/src/main/scala/org/apache/spark/graph/impl/Serializers.scala [hunk 1]
```diff
@@ -80,9 +101,9 @@ class IntAggMsgSerializer extends Serializer {
 
     override def serializeStream(s: OutputStream) = new ShuffleSerializationStream(s) {
       def writeObject[T](t: T) = {
-        val msg = t.asInstanceOf[AggregationMsg[Int]]
-        writeLong(msg.vid)
-        writeUnsignedVarInt(msg.data)
+        val msg = t.asInstanceOf[(Vid, Int)]
+        writeLong(msg._1)
```
Comment (ankurdave) [inline graph/src/main/scala/org/apache/spark/graph/impl/Serializers.scala:105]: @rxin Why doesn't IntAggMsgSerializer use writeVarLong for the vid like the others do?

Comment (rxin) [inline graph/src/main/scala/org/apache/spark/graph/impl/Serializers.scala:105]: Good idea. Can you make that change?

Comment (ankurdave) [inline graph/src/main/scala/org/apache/spark/graph/impl/Serializers.scala:105]: Done as #85.

**Pull request merged**